pumps/resurface.go

```diff
@@ -0,0 +1,234 @@
+		"API-ID",
+		"API-Key",
+		"API-Name",
+		"API-Version",
+		"Oauth-ID",
+		"Org-ID",
+	}
+
+	// Decode raw HTTP transaction from base64 strings
+	rawBytesReq, err := base64.StdEncoding.DecodeString(rec.RawRequest)
+	if err != nil {
+		return
+	}
+	rawBytesRes, err := base64.StdEncoding.DecodeString(rec.RawResponse)
+	if err != nil {
+		return
+	}
+	rawReq := string(rawBytesReq)
+	rawRes := string(rawBytesRes)
+
+	// Slice first line, headers, body+trailers
+	copy(req[:2], strings.SplitN(rawReq, "\r\n", 2))
+	copy(res[:2], strings.SplitN(rawRes, "\r\n", 2))
+	copy(req[1:], strings.SplitN(req[1], "\r\n\r\n", 2))
+	copy(res[1:], strings.SplitN(res[1], "\r\n\r\n", 2))
+
+	// Request method
+	method := rec.Method
+	if method == "" {
+		method = strings.Fields(req[0])[0]
+	}
+
+	// Request URL
+	// schema := "http" // TODO - could the AnalyticsRecord struct be modified to include the target URL Schema?
+	path := rec.RawPath
+	rawPath := strings.Fields(req[0])[1]
+	if path == "" {
+		path = rawPath
+	} else if idx := strings.Index(rawPath, "?"); idx != -1 {
+		path = path + rawPath[idx:]
+	}
+	if !strings.Contains(path, "://") && !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	parsedURL, err := url.Parse(path)
+	if err != nil {
+		return
+	}
+
+	// Request headers
+	reqHeaders := parseHeaders(req[1], nil)
+
+	// Request address
+	if reqHeaders.Get("X-FORWARDED-FOR") == "" {
+		reqHeaders.Add("X-FORWARDED-FOR", rec.IPAddress)
+	}
+
+	// Request host
+	host := rec.Host
+	if host == "" {
+		host = reqHeaders.Get("Host")
+	}
+
+	// Custom Tyk fields
+	customFields = make(map[string]string, len(tykFields))
+	for _, field := range tykFields {
+		key := strings.ReplaceAll(field, "-", "")
+		if value := reflect.ValueOf(&rec).Elem().FieldByName(key).String(); value != "" {
+			customFields["tyk-"+field] = value
+		}
+	}
+
+	// Response Status
+	status := rec.ResponseCode
+	if status == 0 {
+		status, err = strconv.Atoi(strings.Fields(res[0])[1])
+		if err != nil {
+			return
+		}
+	}
+
+	// Response Headers
+	resHeaders := parseHeaders(res[1], nil)
+
+	// Response Trailers
+	if res[2] != "" && resHeaders.Get("Transfer-Encoding") == "chunked" && resHeaders.Get("Trailer") != "" {
+		lastChunkIndex := strings.LastIndex(res[2], "0\r\n") + 3
+		resHeaders = parseHeaders(res[2][lastChunkIndex:], resHeaders)
+		res[2] = res[2][:lastChunkIndex]
+	}
+
+	httpReq = http.Request{
+		Method: method,
+		Host:   host,
+		URL:    parsedURL,
+		Header: reqHeaders,
+		Body:   ioutil.NopCloser(strings.NewReader(req[2])),
+	}
+
+	if parsedURL.IsAbs() {
+		httpReq.RequestURI = path
+	}
+
+	httpResp = http.Response{
+		StatusCode: status,
+		Header:     resHeaders,
+		Body:       ioutil.NopCloser(strings.NewReader(res[2])),
+	}
+
+	return
+}
+
+func (p *ResurfacePump) WriteData(ctx context.Context, data []interface{}) error {
+	p.log.Debug("Writing ", len(data), " records")
+
+	for _, v := range data {
+		decoded := v.(analytics.AnalyticsRecord)
```

Comment:
It would be good to check the casting here. 
Something like this should do the trick:
```go
decoded, ok := v.(analytics.AnalyticsRecord)
if !ok {
   p.log.Error("error decoding analytic record")
   continue
}
```